### PR TITLE
Device credentials enhancement 

### PR DESCRIPTION
**Pull request by Kandarimegha**
## Description:  The enhancement of the Device Credential Workflow Manager has been successfully completed with the following improvements:
Added API to sync the device credentials to the site device, in case device/s is/are present inside the site. 
Uplifted the deprecating APIs with the new APIs.
Version-Based Routing: Implemented version-based routing for the required methods.
                           
Please include a summary of the changes and the related issue. Also, include relevant motivation and context.

## Type of Change
- [ No] Bug fix
- [ Yes] New feature
- [ No] Breaking change
- [ Yes] Documentation update

## Checklist
- [ Yes] My code follows the style guidelines of this project
- [ Yes] I have performed a self-review of my own code
- [ Yes] I have commented my code, particularly in hard-to-understand areas
- [ Yes] I have made corresponding changes to the documentation
- [Yes ] My changes generate no new warnings
- [No ] I have added tests that prove my fix is effective or that my feature works
- [No ] New and existing unit tests pass locally with my changes
- [No ] Any dependent changes have been merged and published in downstream modules
- [No ] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [ Yes] Tasks are idempotent (can be run multiple times without changing state)
- [No ] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [ Yes] Playbooks are modular and reusable
- [ No] Handlers are used for actions that need to run on change

## Documentation
- [Yes ] All options and parameters are documented clearly.
- [Yes ] Examples are provided and tested.
- [ Yes] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/device_credential_workflow_manager.py [hunk 1]
```diff
@@ -919,14 +1039,18 @@ def get_global_credentials_params(self):
...
-            return None
-
+        except Exception as msg:
+            self.msg = (
+                "Exception occurred while getting global device credentials: {0}".format(
+                    msg)
+            )
+            self.log(self.msg, "CRITICAL")
+            self.status = "failed"
+            return self.check_return_status()
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py:1075]: Done! added a line after the try: catch: block.

File: plugins/modules/device_credential_workflow_manager.py [hunk 2]
```diff
@@ -2359,6 +2772,214 @@ def assign_credentials_to_site(self):
...
+        return site_credential_response
+
+    def apply_credentials_to_site(self):
+        """
+        Apply Global Device Credential to the Cisco Catalyst
+        Center based on the provided playbook details.
+        Check the return value of the API with check_return_status().
+
+        Parameters:
+            self
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Added self -definition to the method description for clarity.

File: plugins/modules/device_credential_workflow_manager.py [hunk 3]
```diff
@@ -2359,6 +2772,214 @@ def assign_credentials_to_site(self):
...
+
+        Returns:
+            self
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Added self -definition to the method description for clarity.

File: plugins/modules/device_credential_workflow_manager.py [hunk 4]
```diff
@@ -2359,6 +2772,214 @@ def assign_credentials_to_site(self):
         self.status = "success"
         return self
 
+    def get_network_devices_credentials_sync_status(self):
+        """
+        Get network devices credentials sync status from Cisco Catalyst Center.
+
+        Parameters:
+            self
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Added self -definition to the method description for clarity.

File: plugins/modules/device_credential_workflow_manager.py [hunk 5]
```diff
@@ -875,27 +982,40 @@ def get_site_id(self, site_name):
...
+
             if not response:
                 self.log("Failed to retrieve the site ID for the site name: {0}"
                          .format(site_name), "ERROR")
                 return None
 
             response = response.get("response")
             if not response:
-                self.log("The site with the name '{0}' is not valid".format(site_name), "ERROR")
+                self.log("The site with the name '{0}' is not valid".format(
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Removed the enter key and checked for sanity errors,  there's no sanity issue.

File: plugins/modules/device_credential_workflow_manager.py [hunk 6]
```diff
@@ -1564,13 +1714,15 @@ def get_want_device_credentials(self, CredentialDetails):
...
-                    update_credential = want.get("want_update").get("cliCredential")
+                    update_credential = want.get(
+                        "want_update").get("cliCredential")
                     update_credential.append({})
                     if item.get("password"):
                         update_credential[update_cli_ptr] \
                             .update({"password": item.get("password")})
                     else:
-                        self.msg = "password is mandatory for updating cli_credential " + str(have_cli_ptr)
+                        self.msg = "password is mandatory for updating cli_credential " + \
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: done the suggested modification.

File: plugins/modules/device_credential_workflow_manager.py [hunk 7]
```diff
@@ -669,6 +741,33 @@
...
+    dnac_version: "{{dnac_version}}"
+    dnac_debug: "{{ dnac_debug }}"
+    dnac_log_level: "{{ dnac_log_level }}"
+    dnac_log: True
+    state: merged
+    config_verify: True
+    config:
+      - apply_credentials_to_site:
+        cli_credential:
+          description: string
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Added actual values for every example.

File: plugins/modules/device_credential_workflow_manager.py [hunk 8]
```diff
@@ -2152,13 +2271,171 @@ def get_want_assign_credentials(self, AssignCredentials):
...
+            that need to be applied to a site from the playbook.
+        """
+        want = {
+            "apply_credentials": {}
+        }
+        site_name = ApplyCredentials.get("site_name")
+        if not site_name:
+            self.msg = "The 'site_name' is required parameter for 'apply_credentials_to_site'"
+            self.status = "failed"
+            return self
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py:2304]: Done

File: plugins/modules/device_credential_workflow_manager.py [hunk 9]
```diff
@@ -2152,13 +2271,171 @@ def get_want_assign_credentials(self, AssignCredentials):
...
+                        return self
+                elif cliDescription and cliUsername:
+                    for item in cli_details:
+                        if item.get("description") == cliDescription and \
+                                item.get("username") == cliUsername:
+                            cliDetail = item
+                    if not cliDetail:
+                        self.msg = "The username and description of the CLI credential are invalid"
+                        self.status = "failed"
+                        return self
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py:2336]: Done

File: plugins/modules/device_credential_workflow_manager.py [hunk 10]
```diff
@@ -2359,6 +2657,224 @@ def assign_credentials_to_site(self):
...
+
+        Parameters:
+            self - The current object with updated Global Device Credential information.
+            site_id - The ID of the site to check for assigned devices.
+
+        Returns:
+            site_response - The response from the API call.
+        """
+        self.log(
+            "Checking assigned devices for site with ID: {0}".format(site_id), "DEBUG")
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Done the change.

File: plugins/modules/device_credential_workflow_manager.py [hunk 11]
```diff
@@ -2359,6 +2657,224 @@ def assign_credentials_to_site(self):
...
+        self.log("Received API response from 'get_site_assigned_network_devices': {0}"
+                 .format(str(site_assigned_network_devices_response)), "DEBUG")
+        site_response = site_assigned_network_devices_response.get(
+            "response", [])
+
+        return site_response
+
+    def get_assigned_device_credential(self, site_id):
+        """
+        Get device credential settings for a site from Cisco Catalyst Center.
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Modified it as "Retrieve device credential configurations for a site from Cisco Catalyst Center."

File: plugins/modules/device_credential_workflow_manager.py [hunk 12]
```diff
@@ -2359,6 +2657,224 @@ def assign_credentials_to_site(self):
...
+        Check the return value of the API with check_return_status().
+
+        Parameters:
+            self - The current object with updated Global Device Credential information.
+
+        Returns:
+            self - The current object with updated Global Device Credential information.
+
+        """
+        if self.dnac_version >= self.version_2_3_7_6:
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Done

File: plugins/modules/device_credential_workflow_manager.py [hunk 13]
```diff
@@ -2359,6 +2657,224 @@ def assign_credentials_to_site(self):
...
+            if not site_response:
+                result_apply_credential.update({
+                    "No Apply Credentials": {
+                        "response": "No Response",
+                        "msg": "No device available in the site"
+                    }
+                })
+                self.msg = "No device available in the site: {0}".format(site_id)
+                self.log(self.msg, "WARNING")
+                self.status = "skipped"
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: changed skipped to "exited".

File: plugins/modules/device_credential_workflow_manager.py [hunk 14]
```diff
@@ -2359,6 +2657,224 @@ def assign_credentials_to_site(self):
...
+
+            valid_sync_cred_ids, invalid_sync_cred_ids = [], []
+
+            for id in not_synced_ids:
+                if id in assigned_site_ids:
+                    valid_sync_cred_ids.append(id)
+                else:
+                    invalid_sync_cred_ids.append(id)
+
+            self.log("Credential IDs {} not assigned to site, so Sync not possible." .format(
```
[the remark this answers is not in the document]
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Done!

File: plugins/modules/device_credential_workflow_manager.py [hunk 15]
```diff
@@ -7,23 +7,25 @@
...
 module: device_credential_workflow_manager
 short_description: Resource module for Global Device Credentials and Assigning Credentials to sites.
 description:
-- Manage operations on Global Device Credentials and Assigning Credentials to sites.
+- Manage operations on Global Device Credentials, Assigning Credentials to sites and Sync Credentials to site device.
 - API to create global device credentials.
 - API to update global device credentials.
 - API to delete global device credentials.
 - API to assign the device credential to the site.
+- API to sync the device credential to the site. Sync functionality will be applicable for DNAC version 2.3.7.6 onwards.
```
Comment (madhansansel) [inline plugins/modules/device_credential_workflow_manager.py:22]: is applicable for Catalyst Center version 2.3.7.6 and later.

File: plugins/modules/device_credential_workflow_manager.py [hunk 16]
```diff
@@ -344,49 +416,49 @@
...
-          username: string
-          password: string
+        - description: HTTP Write1
+          username: HTTP Write1
+          password: '12345'
           port: 443
       assign_credentials_to_site:
         cli_credential:
-          id: string
+          id: '1b3777a0-09c2-488b-a64c-6f3c441e343'
```
Comment (madhansansel) [inline plugins/modules/device_credential_workflow_manager.py:449]: Why do we need to pass id here? IDs are not easy to remember. Can you give description?

File: plugins/modules/device_credential_workflow_manager.py [hunk 17]
```diff
@@ -470,40 +542,40 @@
...
-        - description: string
-          username: string
-          password: string
-          enable_password: string
-          id: string
+        - description: CLI1
+          username: cli1
+          password: '12345'
+          enable_password: '12345'
+          id: '1b3777a0-09c2-488b-a64c-6f3c441e343'
```
Comment (madhansansel) [inline plugins/modules/device_credential_workflow_manager.py:549]: id is not required..

File: plugins/modules/device_credential_workflow_manager.py [hunk 18]
```diff
@@ -297,6 +299,57 @@
               id:
                 description: snmp_v3 Credential Id. Use Description or Id.
                 type: str
+      apply_credentials_to_site:
```
Comment (madhansansel) [inline plugins/modules/device_credential_workflow_manager.py:302]: Why do we need the sub-section "apply_credentials_to_site"? I only see that the 'site_name' is different. Can't we give the site_name when you create/update credentials? If the site_name is there, then create the credential and apply it to the site. If it is not there, then ignore it.

File: plugins/modules/device_credential_workflow_manager.py [hunk 19]
```diff
@@ -2003,7 +2094,12 @@ def get_want_assign_credentials(self, AssignCredentials):
                         self.msg = "The username and description of the CLI credential are invalid"
                         self.status = "failed"
                         return self
-                want.get("assign_credentials").update({"cliId": cliDetail.get("id")})
+                if self.get_ccc_version_as_integer() <= self.get_ccc_version_as_int_from_str("2.3.5.3"):
```
Comment (madhansansel) [inline plugins/modules/device_credential_workflow_manager.py:2097]: if it is used in multiple places, then 
current_ccc_version_as_int = self.get_ccc_version_as_integer()

File: plugins/modules/device_credential_workflow_manager.py (hunk 15 above)
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Modified this.

File: plugins/modules/device_credential_workflow_manager.py (hunk 16 above)
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: removed id and added description.

File: plugins/modules/device_credential_workflow_manager.py (hunk 17 above)
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: removed.

File: plugins/modules/device_credential_workflow_manager.py (hunk 18 above)
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: Removed apply_credentials_to_site. introduced a new variable "Sync " inside assigned_credentials_to_site, which is of boolean type.

File: plugins/modules/device_credential_workflow_manager.py (hunk 19 above)
Comment (Kandarimegha) [reply plugins/modules/device_credential_workflow_manager.py]: added current_ccc_version_as_int = self.get_ccc_version_as_integer() in def get_want_assign_credentials(self, AssignCredentials) where it is used multiple times

File: plugins/modules/device_credential_workflow_manager.py [hunk 20]
```diff
@@ -1124,24 +1224,7 @@ def get_cli_credentials(self, CredentialDetails, global_credentials):
...
-                                self.status = "failed"
-                                return self
-                            cliDetail = item
-                    if not cliDetail:
-                        self.msg = "CLI credential old_description or old_username is invalid"
-                        self.status = "failed"
-                        return self
+                        return self.check_return_status()
 
                 cliDescription = cliCredential.get("description")
```
Comment (madhansansel) [inline plugins/modules/device_credential_workflow_manager.py:1229]: can you change all with snake case?
cli_description
cli_username 
cli_detail

File: plugins/modules/device_credential_workflow_manager.py [hunk 21]
```diff
@@ -2152,13 +2290,183 @@ def get_want_assign_credentials(self, AssignCredentials):
...
+                self.status = "failed"
+                return self
+            site_ids.append(current_site_id)
+
+        want.update({"site_id": site_ids})
+        want.update({"site_name": site_names})
+        global_credentials = self.get_global_credentials_params()
+        cli_credential = ApplyCredentials.get("cli_credential")
+        if cli_credential:
+            cliId = cli_credential.get("id")
```
Comment (madhansansel) [inline plugins/modules/device_credential_workflow_manager.py:2342]: cli_id
cli_description
cli_username

**Review by madhansansel — COMMENTED** [review]
Most variables and function names are written in Camel case. We need to change them into snake case. Can you start working on it?

**Pull request merged**